File: 10_16_2016_search-for-a-range/chang.java
public class Solution {
    /** 
     *@param A : an integer sorted array
     *@param target :  an integer to be inserted
     *return : a list of length 2, [index1, index2]
     */
    public int[] searchRange(int[] A, int target) {
        // write your code here
        int[] bound = new int[]{-1, -1};
        if (A == null || A.length == 0) {
            return bound;
        }
        //search for the left bound
        int start = 0;
        int end = A.length - 1;
        
        while (start + 1 < end) {
            int mid = start + (end - start) / 2;
            if (A[mid] == target) {
                end = mid;
            }else if (A[mid] < target) {
                start = mid;
            }else {
                end = mid;
            }
        }
        
        if (A[start] == target) {
            bound[0] = start;
        }else if (A[end] == target) {
            bound[0] = end;
        }
        
        //search for the right bound
        start = 0;
        end = A.length - 1;
        
        while (start + 1 < end) {
            int mid = start + (end - start) / 2;
            if (A[mid] == target) {
                start = mid;
            }else if (A[mid] < target) {
                start = mid;
            }else {
                end = mid;
            }
        }
        
        if (A[end] == target) {
            bound[1] = end;
        }else if (A[start] == target) {
            bound[1] = start;
        }
        
        return bound;

        
    }
}

File: 12_01_2016_unique_paths/chang.py
class Solution:
    """
    @param n and m: positive integer(1 <= n , m <= 100)
    @return an integer
    """ 
    def uniquePaths(self, m, n):
        if m == 0 or n == 0:
            return 0
            
        A = [[1 for x in range(n)] for x in range(m)]
            
        '''for i in range(0, m):
            A[i][0] = 1
        for i in range(0, n):
            A[0][i] = 1'''
        
        for i in range(1, m):
            for j in range(1, n):
                A[i][j] = A[i-1][j] + A[i][j-1]
                
        return A[-1][-1]
